### ISSUE TITLE
Multioutput Regression

### ISSUE THREAD

**Issue by ptyshevs** [online-ml/river#433]
```python
from river import compose
from river import linear_model
from river import preprocessing

model = compose.Pipeline(
     preprocessing.MinMaxScaler(),
     linear_model.LinearRegression()
 )
```

This model seems to support only a single-output `y`. Is there any way to make it more general?

As far as I understood, current only option will be to use `river.multioutput.RegressionChain`, which is not suitable in my case.

**Comment by MaxHalford** [comment]
Hello!

Could you please provide more detail as to why `RegressorChain` is not suitable?

**Comment by ptyshevs** [comment]
@MaxHalford  the only thing that bothers me is that for each output previous output is used as a feature.

Given m features at time t, I am building a model to predict features at time t+1. Hence, there are m inputs and m outputs.

**Comment by MaxHalford** [comment]
I still don't see why you can't use RegressorChain. Maybe you could provide a more detailed example and point out what's failing? In any case, RegressorChain is the only multioutput model for regression.

**Comment by ptyshevs** [comment]
@MaxHalford oh, sorry, I misunderstood your question. It is not that I cannot use RegressorChain -- it works fine. It's more that I don't want the specific chaining logic to happen, since prediction of feature X_1 should not go as input to predictor of feature X_2.

If there is no other option, it's totally fine, I just need to confirm that that's the case.

**Comment by smastelini** [comment]
Hey everyone, I'm a bit late on this discussion. 

@ptyshevs, you can also rely on `tree.iSOUPTreeRegressor` for multi-target regression. 

In the future we might add a `MultiOutputRegressor` wrapper (as in sklearn and in the legacy skmultiflow) to create a independent model per target. 

Cheers!

**Comment by eserie** [comment]
Hi everyone,
would you consider to implement a multi-output regression based on a statistical method such as the minimization of the negative log-likelihood of a multi-variate Gaussian distribution? The approach is described in [1] (along with other "problem transformation methods" such as Chain regression). An online version should should be easy to formulate. 

[1] Borchani, H., Varando, G., Bielza, C., & Larranaga, P. (2015). A survey on multi‐output regression. Wiley Interdisciplinary Reviews: Data Mining and Knowledge Discovery, 5(5), 216-233.

**Comment by MaxHalford** [comment]
To be totally we're open to everything as long it has a decent amount of scientific "support" (the paper you mention is well enough cited). We also only want to include methods which are reasonably efficient.

**Comment by eserie** [comment]
Hi @MaxHalford , thanks for the quick reply!
To be honest, I recently made the exercise to implement a version within river and it seems to fit well with the API. In terms of speed it is possible to achieve good performance compared to chain regression, depending of the learning scheme you choose to learn the covariance / precision matrix. Also I played with full numpy implementation. Do you mind to use dict of float format or using numpy should be ok? 
In [1] they describe the method as one of the first introduced method for multi-output regressions (in a batch setting), but I could not find any reference for an online version.

**Comment by MaxHalford** [comment]
> To be honest, I recently made the exercise to implement a version within river and it seems to fit well with the API.

I think this is always the good way to start: write something scrappy in a notebook and try it.

> Do you mind to use dict of float format or using numpy should be ok?

We really prefer that everything works with dicts. It's rare that numpy is faster than Python dicts when you work with single samples. However, in some cases numpy is unavoidable due to linear algebra routines that are too painful to write with dicts, in which case we make exceptions. Regardless, the user-facing API is and always will be dicts. When scikit-multiflow merged with creme and adopted creme's API design, they added [some utilities](https://github.com/online-ml/river/blob/master/river/utils/data_conversion.py) to convert numpy arrays to dicts.

> In [1] they describe the method as one of the first introduced method for multi-output regressions (in a batch setting), but I could not find any reference for an online version.

Yes I've often encountered the same situations. Some people don't even realise that their stuff can be adapted online.

**Comment by ptyshevs** [comment]
@eserie , @MaxHalford  maybe I'm missing something, but how can negative log-likelihood minimization be helpful in regression problem? I was thinking about just averaging L2 loss of each output and then doing back-propagation, similarly to a shallow neural net. Tell me what you think.

**Comment by MaxHalford** [comment]
Log-likelihood minimization isn't limited to classification problems, it's just a way to formulate a problem. 

> I was thinking about just averaging L2 loss of each output and then doing back-propagation, similarly to a shallow neural net.

Yes that's called softmax regression for classification. I guess you could imagine some sort of averaged loss function for regression!

**Comment by eserie** [comment]
For regression problem, a statistical approach can be seen as a way to specify correlation structure between outputs. In case where the covariance estimate is fixed (should not be learned from the data), it can be viewed as a [generalized least square](https://en.wikipedia.org/wiki/Generalized_least_squares) or [weighted least square](https://en.wikipedia.org/wiki/Weighted_least_squares). The statistical formulation could permit to consider situations where you learn also the target correlation structure inline.
If you mind, I can try to make a proof of concept in a notebook and share with you to see if we can go further in river.

**Comment by MaxHalford** [comment]
Yep, sounds good :)

**Comment by eserie** [comment]
Hi, I implemented a version of StatisticalRegression in [this notebook](https://github.com/eserie/river_contribs/blob/master/01_Stat_model.ipynb). Sorry the notebook mixes with some other things but I could clean it a bit if needed (In particular jax part is quite optional and may be not compatible with river API).
